### ISSUE TITLE
Support conditional dependencies based on the current build configuration

### ISSUE THREAD

**Issue by tonyarnold** [apple/swift-package-manager#5645]
### Description

My product vends an XCFramework that we would like to only link during debug builds (it is a debugging tool). 

### Expected behavior

I can specify that a target that is included or excluded based upon the build configuration, ie:

```swift
// Would only build and link the target when building with a Debug configuration
.target(name: "RevealUI", condition: .when(configurations: [.debug]))
```

### Actual behavior

Currently, it appears that there is support for conditionally including targets and dependencies based on the platform (shown below), but there is no support for build configuration conditions.

```swift
.target(name: "RevealUI", condition: .when([.macOS, .iOS]))
```

### Swift Package Manager version/commit hash

5.7

### Swift & OS version (output of `swift --version && uname -a`)

swift-driver version: 1.60 Apple Swift version 5.7 (swiftlang-5.7.0.120.1 clang-1400.0.28.1)
Target: arm64-apple-macosx13.0
Darwin Tonys-MBP.localdomain 22.0.0 Darwin Kernel Version 22.0.0: Tue Jun 28 20:46:36 PDT 2022; root:xnu-8792.0.134.131.2~1/RELEASE_ARM64_T6000 arm64

**Comment by ahmetgeymen** [comment]
Friendly reminder that this topic was brought up again on the Swift Forums in March 2026:

https://forums.swift.org/t/se-0273-configuration-conditionals-for-target-dependencies-time-to-finish-the-job/85669

**Later edit:** This topic was brought up again by the same developer (@dmhts) on the forums:

https://forums.swift.org/t/pitch-swiftpm-build-configuration-conditionals-for-target-dependencies/88981 (August 2026)

**Comment by ahmetgeymen** [comment]
### Workaround: Handling Multiple Environment Binaries (.xcframework) in Local SPM

If you are struggling with Swift Package Manager's lack of support for conditional binary targets based on build configurations (e.g., `Debug` vs `Release`), here is a clean workaround using an **Xcode Scheme Pre-Action script** and a **dynamic symlink**.

This approach allows you to maintain a single app target, avoid modifying the binary frameworks themselves, and ensure only the correct environment binary is bundled into your final build.

> Note: This approach and the script below are based on using the framework as a local Swift Package Manager (SPM) dependency.

---

#### Step 1: Configure `Package.swift` to point to a dynamic symlink
Instead of hardcoding separate targets for each environment, point your binary target to an active symlink wrapper name (e.g., `my-framework-active.xcframework`):

```swift
// swift-tools-version: 5.9
import PackageDescription

let package = Package(
    name: "MyFramework",
    products: [
        .library(
            name: "MyFramework",
            targets: ["MyFrameworkTarget"]
        )
    ],
    targets: [
        .binaryTarget(
            name: "MyFrameworkTarget",
            path: "my-framework-active.xcframework" // Symlink managed dynamically by Xcode scheme pre-actions
        )
    ]
)
```

#### Step 2: Add an Xcode Scheme Pre-Action Script

Configure a **Pre-action** in your Xcode scheme to evaluate the active build configuration (`${CONFIGURATION}`) and dynamically redirect the symlink before each build.

1. Open **Edit Scheme...**
2. Select **Build → Pre-actions**.
3. Click **New Run Script Action**.
4. Make sure **Provide build settings from** is set to your **app target**.
5. Add the following script:

```sh
# Path to your local SPM package directory relative to the project
PACKAGE_DIR="${SRCROOT}/../MyFrameworkPackage"

echo "Current Build Configuration: ${CONFIGURATION}"

# Navigate to the package directory to ensure relative symlinks resolve correctly
cd "${PACKAGE_DIR}" || exit 1

# Handle environment-specific framework redirection via 'my-framework-active.xcframework'
if [[ "${CONFIGURATION}" == *"Debug"* ]]; then
    echo "Pointing framework to Develop variant..."
    ln -sfn "my-framework-develop.xcframework" "my-framework-active.xcframework"
else
    echo "Pointing framework to Production variant (Release)..."
    ln -sfn "my-framework-release.xcframework" "my-framework-active.xcframework"
fi

```

**Comment by dschaefer2** [comment]
This is a pretty ugly hack. Good news is that we do have conditionals on targets on our roadmap, at least platform conditionals, though I could see configuration conditionals working too. We just need someone to work on it.

**Comment by tonyarnold** [comment]
> We just need someone to work on it.

How can I help? Is this something someone from outside of your team can assist with?

**Comment by dschaefer2** [comment]
@tonyarnold Definitely. I'm not sure there's anything blocking it. @owenv may know more. I know our platform conditionals are lacking in precision at the moment. (e.g. which `.linux`?). I think the big challenge is how does this come across the PIF and into SwiftBuild so that it knows when to not build a target. Or is this get applied to the dependencies which already support platform conditions, but not configuration conditions.

**Comment by owenv** [comment]
https://github.com/swiftlang/swift-evolution/pull/3433/changes

**Comment by dmhts** [comment]
I have this already working on my local machine, just need to wrap up some tests. The PR is coming in the next couple of days. The only thing is, the proposal, that Owen referenced above, still needs to go through review. Maybe it could be a shorter one as the proposed change is unusually straightforward.

**Comment by dschaefer2** [comment]
https://github.com/swiftlang/swift-evolution/pull/3433/changes covers configuration conditions on dependencies. This issue was about conditionals on targets themselves. Though I guess that could be factored onto all the dependencies on those targets.

**Comment by tonyarnold** [comment]
What @dmhts is doing _seems_ to be what I need, so perhaps my issue was not as well formed as I'd hoped. Conditional dependencies is absolutely what I want (with coverage for all of the existing dependency types, including binaryTarget).

@dschaefer2 are you interpreting this as an exposed target only being available to consumers under provided conditions?

**Comment by dschaefer2** [comment]
> @dschaefer2 are you interpreting this as an exposed target only being available to consumers under provided conditions?

I was just reacting to the original example:
```
.target(name: "RevealSDK", condition: .when(configurations: [.debug]))
```
where it appears the conditional is actually on the target, not on dependencies. That's something others have requested as well. But if you want to focus just on dependencies for now, that's a reasonable first step.

**Comment by dmhts** [comment]
Here you go: https://github.com/swiftlang/swift-package-manager/pull/10529
